core/src/main/java/org/opensearch/sql/planner/physical/LookupOperator.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.sql.planner.physical;
+
+import static org.opensearch.sql.data.type.ExprCoreType.STRUCT;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiFunction;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.ToString;
+import org.opensearch.sql.data.model.ExprTupleValue;
+import org.opensearch.sql.data.model.ExprValue;
+import org.opensearch.sql.data.model.ExprValueUtils;
+import org.opensearch.sql.expression.ReferenceExpression;
+
+/** Lookup operator. Perform lookup on another OpenSearch index and enrich the results. */
+@Getter
+@ToString
+@EqualsAndHashCode(callSuper = false)
+public class LookupOperator extends PhysicalPlan {
+  @Getter private final PhysicalPlan input;
+  @Getter private final String indexName;
+  @Getter private final Map<ReferenceExpression, ReferenceExpression> matchFieldMap;
+  @Getter private final Map<ReferenceExpression, ReferenceExpression> copyFieldMap;
```

Comment:
suggest using [off-heap map](https://github.com/apache/datasketches-memory) for this high memory use case